# Tugas Pendahuluan Modul 5 PBF

### Nama : Rafi Fadhlillah
### NIM : 121450143
### Kelas : RC

### 1. Buatlah sebuah decorator hitung_waktu dan membuat fungsi perkalian yang menerima dua parameter bilangan bulat hingga mengembalikan hasil perkalian dari kedua bilangan tersebut.
a. Isi code pemorgraman di bawah ini dengan menghasilkan output sebagai berikut: <br>
b. Gunakan fungsi wrapper untuk menjalankan fungsi f didalam 'hitung_waktu'

In [1]:
import time 
def hitung_waktu(f):
    def wrapper(a, b):
        start = time.time()
        hasil = f(a, b)
        end = time.time()
        print(f"Waktu yang diperlukan: {end - start} detik")
        return hasil
    return wrapper

def perkalian(x, y):
    #Buatlah Fungsi yang menghitung perkalian dua bilangan bulat
    return x * y

hasil_perkalian = perkalian(5, 7)
hasil_hitung_waktu = hitung_waktu(perkalian)(5, 7)

print(hasil_perkalian)
print(hasil_hitung_waktu)

Waktu yang diperlukan: 0.0 detik
35
35


Catatan : output Waktu yang diperlukan berbeda-beda, hal ini disesuaikan dengan sinyal/jaringan pada device masing masing

### 2. Analisis kode fungsi komposisi dibawah ini!

In [2]:
def compose(g, f):
    def h(*args, **kwargs):
        return g(f(*args, **kwargs))
    return h

celcius_to_fahrenheit = lambda x: (x*9/5) + 32
fahrenheit_to_kelvin = lambda x: ((x-32) * 5/9) + 273
celcius_to_kelvin = compose(fahrenheit_to_kelvin, celcius_to_fahrenheit)
celcius_to_kelvin(25)

298.0

#### a. Jelaskan alur program kode diatas untuk mencapai hasil 
Program di atas merupakan contoh implementasi dari konsep komposisi fungsi pada bahasa pemrograman Python. Pada baris pertama dan kedua, didefinisikan dua fungsi lambda yaitu celcius_to_fahrenheit dan fahrenheit_to_kelvin yang berfungsi untuk mengonversi suhu dari Celcius ke Fahrenheit dan dari Fahrenheit ke Kelvin. Pada baris keempat, fungsi compose didefinisikan dengan dua parameter yaitu g dan f, yang masing-masing merepresentasikan fungsi fahrenheit_to_kelvin dan celcius_to_fahrenheit. Fungsi compose mengembalikan fungsi baru h, yang memiliki parameter `*args` dan `**kwargs.` Fungsi h akan mengembalikan nilai dari komposisi fungsi g dan f pada parameter `*args` dan `**kwargs.` <br>

Pada baris terakhir, dilakukan penggunaan fungsi compose untuk mendefinisikan fungsi baru celcius_to_kelvin. Fungsi ini merupakan hasil komposisi fungsi fahrenheit_to_kelvin dan celcius_to_fahrenheit. Dalam hal ini, f adalah celcius_to_fahrenheit dan g adalah fahrenheit_to_kelvin. Kemudian, pada baris terakhir dilakukan pemanggilan fungsi celcius_to_kelvin dengan argumen 25. Hasil yang diharapkan adalah konversi suhu dari 25 derajat Celcius ke Kelvin.<br>

#### b. Jika g(f(*args, **kwargs)) diubah f(g(*args, **kwargs)) apakah hasil tetap 298 Kelvin? Jelaskan menurut pendapat anda

Jika `g(f(*args, **kwargs))` diubah menjadi `f(g(*args, **kwargs))`, maka hasil yang diperoleh tidak akan sama dengan 298 Kelvin. Hal ini disebabkan karena urutan komposisi fungsi akan berubah. Pada komposisi awal, fungsi celcius_to_fahrenheit digunakan terlebih dahulu untuk mengonversi suhu dari Celcius ke Fahrenheit, kemudian hasilnya digunakan oleh fungsi fahrenheit_to_kelvin untuk mengonversi suhu dari Fahrenheit ke Kelvin. Sedangkan pada komposisi yang diubah, fungsi fahrenheit_to_kelvin digunakan terlebih dahulu untuk mengonversi suhu dari Fahrenheit ke Kelvin, kemudian hasilnya digunakan oleh fungsi celcius_to_fahrenheit untuk mengonversi suhu dari Celcius ke Fahrenheit. Karena adanya perbedaan urutan konversi suhu ini, maka hasil yang diperoleh akan berbeda pula.

### 3. perhatikan code dibawah ini!

In [3]:
def my_function():
    func =[]
    def luas(p, l):
        return p*l
    func.append(luas)
    def keliling(p, l):
        return 2*(p+l)
    func.append(keliling)
    user = input('luas/keliling : ')
    if user.lower() == "luas":
        return func [0] (1,2)
    elif user.lower() == "keliling":
        return func [1](3,2)
    else:
        print("Input yang dimasukkan tidak valid. Silakan coba lagi.")
        return user
    return func

result = my_function()
print("Hasil  :",result)

luas/keliling :  keliling


Hasil  : 10


#### a. Jelaskan konsep closure dari code! <br>
Konsep closure pada kode di atas adalah dimana terdapat sebuah fungsi my_function() yang mengandung dua fungsi nested (fungsi dalam fungsi) yaitu luas() dan keliling(). Fungsi-fungsi tersebut ditambahkan ke dalam sebuah list bernama func menggunakan metode append(). Kemudian, fungsi my_function() mengambil input dari pengguna untuk memilih apakah ingin menghitung luas atau keliling, dan kemudian memanggil fungsi luas() atau keliling() yang disimpan di dalam list func. Fungsi my_function() mengembalikan nilai dari fungsi luas() atau keliling() tergantung pada pilihan pengguna. Closure terjadi karena fungsi luas() dan keliling() dapat diakses di dalam fungsi my_function() meskipun kedua fungsi tersebut tidak di definisikan di luar fungsi my_function(). Selain itu, fungsi luas() dan keliling() memiliki akses ke variabel func yang didefinisikan di dalam fungsi my_function(), yang membuat mereka menjadi closure. 

#### b. Eksekusi code diatas kemudian jelaskan mengapa jika menginput "luas" hasil yang keluar adalah 2 dan jika menginput keliling hasilnya 10
Jika user menginput "luas", maka program akan mengembalikan hasil dari fungsi luas dengan parameter 1 dan 2. Sehingga hasilnya adalah 2. Sedangkan jika user menginput "keliling", maka program akan mengembalikan hasil dari fungsi keliling dengan parameter 3 dan 2. Sehingga hasilnya adalah 10. Ini karena rumus luas adalah panjang x lebar, yang pada kasus ini adalah 1 x 2 = 2. Sedangkan rumus keliling adalah 2 x (panjang + lebar), yang pada kasus ini adalah 2 x (3 + 2) = 10. Jika user menginput input yang tidak valid, maka program akan mencetak pesan kesalahan dan mengembalikan nilai input tersebut.